### PR TITLE
Vectorize leaf dynamics

### LINKED ISSUE #19: [Tasks]: Make the computations of `leaf_dynamic` vectorized
After merging #12 , we want to change the data array organization inside the module, see #18. For now, focus only on part A and organize arrays as:

- First dimension indexes over data examples: n_data.
- Second dimension indexes over parameter sets: n_param_sets
- Third dimension indexes over locations: n_locations.

Requirements:

- The existing tests should pass. 
- New tests should be implemented for vectorization.

### PR DESCRIPTION
**Pull request by fnattino**
Closes #19

### CODE REVIEW

**Comment by fnattino** [comment]
@SarahAlidoost, I left one question below, but this is otherwise ready to be reviewed. 

I have added few tests that show that the module can now work with arbitrarily shaped arrays as parameters (provided all parameters have the same shape). There is quite some repeated code in the tests, happy to hear whether you have suggestions for improvements there. Otherwise, I think that things will simplify significantly when we will modify the engine class and make it easier to pass input and get output.

File: src/diffwofost/physical_models/crop/leaf_dynamics.py [hunk 1]
```diff
@@ -236,57 +242,48 @@ def calc_rates(self, day, drv):
...
-                weight = torch.sigmoid((lvage - p.SPAN) * sharpness)
-                DALV = DALV + weight * lv
-        else:
-            for lv, lvage in zip(s.LV, s.LVAGE, strict=False):
-                if lvage > p.SPAN:
-                    DALV = DALV + lv
-
-        r.DALV = DALV
+        tSPAN = _broadcast_to(p.SPAN, s.LVAGE.shape)  # Broadcast to same shape
+        sharpness = torch.tensor(1000.0, dtype=DTYPE)  # FIXME
```
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply src/diffwofost/physical_models/crop/leaf_dynamics.py]: @fnattino right! can you please replace it with `torch.where`. Then we dont need approximation using sigmoid.

File: src/diffwofost/physical_models/crop/leaf_dynamics.py [hunk 2]
```diff
@@ -113,14 +112,17 @@ class WOFOST_Leaf_Dynamics(SimulationObject):
     LAI, TWLV
     """
 
+    START_DATE = None
+    MAX_DAYS = 300
```
Comment (SarahAlidoost) [inline src/diffwofost/physical_models/crop/leaf_dynamics.py]: Can you please add a comment explaining these two variables. Thanks.

File: src/diffwofost/physical_models/crop/leaf_dynamics.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (SarahAlidoost) [reply src/diffwofost/physical_models/crop/leaf_dynamics.py]: Good catch! Thanks. As you said, the reason differentiability w.r.t `SPAN` works with the sigmoid but not with `torch.where` is that `torch.where` introduces a hard threshold, just like the original code `if lvage > p.SPAN`. But, the sigmoid is considerd as a soft transition around `SPAN`. So we should keep the sigmoid. We can also add a comment in the code to explain why we’re using a sigmoid instead of `torch.where`. Can you please add a comment?

In the case of using `torch.where` in [these lines](https://github.com/WUR-AI/diffWOFOST/blob/5d28d3ff49e71f070b7046b6dc1a0ee11a2250c7/src/diffwofost/physical_models/crop/leaf_dynamics.py#L278-L286), we don’t yet have tests for differentiability w.r.t the parameter `p.RGRLAI`.  If later we see tests fail, issue will be likely the hard condition `is_lai_exp = s.LAIEXP < 6.0`, which can be replaced with sigmoid. Not all parameters are differentiable at the moment, only `TDWI` and `SPAN`, see tests and issue #8.

In case of using `torch.where` in [these lines](https://github.com/WUR-AI/diffWOFOST/blob/5d28d3ff49e71f070b7046b6dc1a0ee11a2250c7/src/diffwofost/physical_models/crop/leaf_dynamics.py#L312-L316), that is fine, because `TDWI` and `SPAN` are not involved in the conditional logic. And because the numerical tests for those cases are passing, it confirms that the gradients are being computed correctly. For the remaining parameters, we’ll verify this later. I will create a backlog item and link to this discussion.

[the remark this answers is not in the document]
Comment (fnattino) [reply src/diffwofost/physical_models/crop/leaf_dynamics.py]: OK, sounds good, on that we are aware that the model might not be differentiable with respect to some parameters - I will add a comment here and close to the `torch.where` statements related to this! Thanks for the input :)

**Review by fnattino — COMMENTED** [review]
@SarahAlidoost I should have addressed your comments above. When adding tests, I have slightly modified the structure of tests for leaf dynamics to be able to run them with parameters, so it would be easier to add more of these tests. 

Now, one of the thing I did to make this implementation easier, is to have `DiffLeafDynamics` return a dictionary of tensors instead of stacking all the tensors in a larger tensor. I have then ported the same structure to the tests on root dynamics, which unveiled an issue there: the `RD` variable seems not to have the gradient defined (see failing tests, e.g. [here](https://github.com/WUR-AI/diffWOFOST/actions/runs/18751586732/job/53492395666?pr=36#step:5:100)). Earlier, this issue was not caught by the tests, because when stacking `RD` and `TWRT` in a single tensor, a gradient would somehow become available - but now this seems to have emerged. Do you agree with the assesment?

**Comment by fnattino** [comment]
Thanks for all the clarifications (and for bearing with me in this PR :) ). I have followed your suggestion and: 

* "unpacked" some of the parametrizations in the tests, putting back one class per parameter as it was originally set up; 
* put back the check on the gradient being zero for `TWLV` wrt `SPAN`. Note, however, that I have setup the comparison on the gradient calculated from autograd (not the numerical one), since the numerical gradient turned out to be small but not exacly zero (~1.e-5). Not sure why this pops up only now. 
* fixed the root dynamics tests to correct the behaviour for checking on `RD` wrt `TDWI`.

**Review by SarahAlidoost — APPROVED** [review]
@fnattino looks great! Thank you for implementing the vectorization and addressing the comments and suggestions. Ready to merge with only one small change, could you please remove the test `test_gradients_tdwi_rd_root_dynamics_numerical` before merging? 🚀
